### PR TITLE
[hierarchical-compilation][invoke_subgraph] Use tracing context to cache artifacts of dispatch keys

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138031
* #137808
* __->__ #137965
* #138036
* #137538

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: torch/_higher_order_ops/invoke_subgraph.py [hunk 1]
```diff
@@ -203,21 +225,28 @@ def invoke_subgraph_func(ctx, subgraph, identifier, operands):
...
-            )
+        skip_check = False
+        invoke_subgraph_cache = get_invoke_subgraph_cache()
+        if invoke_subgraph_cache:
+            if invoke_subgraph_cache.get_functional_tensor_entry(identifier):
+                skip_check = True
+
+        if not skip_check:
+            pre_dispatch = hasattr(ctx, "mode") and ctx.mode.pre_dispatch
+            if _has_potential_branch_input_mutation(
```
Comment (anijain2305) [inline torch/_higher_order_ops/invoke_subgraph.py:236]: Should be moved to Dynamo. If there is a mutation, don't even create invoke_subgraph call.

**Comment by anijain2305** [comment]
@zou3519 This is ready for review.

File: torch/_guards.py [hunk 2]
```diff
@@ -570,6 +570,66 @@ def restore_graphstate(self, state):
         self.dynamo_guards = GuardsSet(state.dynamo_guards)
 
 
+class HopSubgraphCache:
```
Comment (zou3519) [inline torch/_guards.py:573]: Do you think this will be useful beyond InvokeSubgraph?

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**